src/plonk_recursion.rs

```diff
@@ -47,9 +48,100 @@ impl RecursionPublicInputs {
             self.o_plonk_t.as_slice(),
             self.halo_u_l.as_slice(),
             self.halo_u_r.as_slice(),
+            self.old_proofs.as_slice(),
         ]
         .concat()
     }
+
+    pub fn populate_witness<C: Curve>(
+        &self,
+        witness: &mut PartialWitness<C::BaseField>,
+        values: Proof<C>,
+    ) -> Result<()> {
+        let challs = values.get_challenges()?;
+        witness.set_public_input(self.beta, challs.beta.try_convert()?);
+        witness.set_public_input(self.gamma, challs.gamma.try_convert()?);
+        witness.set_public_input(self.alpha, challs.alpha.try_convert()?);
+        witness.set_public_input(self.zeta, challs.zeta.try_convert()?);
+        witness.set_public_input(
+            self.o_plonk_z_local,
+            values.o_local.o_plonk_z.try_convert()?,
+        );
+        witness.set_public_input(
+            self.o_plonk_z_right,
+            values.o_right.o_plonk_z.try_convert()?,
+        );
+        for (&a, b) in self
+            .o_constants
+            .iter()
+            .chain(&self.o_plonk_sigmas)
+            .chain(&self.o_local_wires)
+            .chain(&self.o_right_wires)
+            .chain(&self.o_below_wires)
+            .chain(&self.o_plonk_t)
+            .chain(&self.halo_u_l)
+            .chain(&self.halo_u_r)
+            .chain(&self.old_proofs)
+            .zip(
+                values
+                    .o_local
+                    .o_constants
+                    .iter()
+                    .chain(&values.o_local.o_plonk_sigmas)
+                    .chain(&values.o_local.o_wires)
+                    .chain(&values.o_right.o_wires)
+                    .chain(&values.o_below.o_wires)
+                    .chain(&values.o_local.o_plonk_t)
+                    .chain(&challs.ipa_challenges)
+                    .chain(&C::ScalarField::batch_multiplicative_inverse(
+                        &challs.ipa_challenges,
+                    )),
+            )
+        {
+            witness.set_public_input(a, b.try_convert()?);
+        }
+
+        Ok(())
+    }
+
+    pub fn proof_to_public_inputs<C: Curve, InnerC: HaloCurve<BaseField = C::ScalarField>>(
```

Comment:
I think it should be possible to do this in a generic way, with public inputs being an output of `generate_proof` (maybe contained in `Proof`). I could try implementing that later.

Reply:
Yes we definitely could. I don't have a strong opinion on this, so if you think it's better to have the PI as an output of `generate_proof` we can implement that in a later PR.